compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -1502,11 +1896,46 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
 
                 let l = self.new_named_lifetime(lifetime.id, id, span, ident);
                 hir::GenericArg::Lifetime(l)
-            }));
+            })
+            .collect();
         debug!(?lifetimes);
 
+        let type_const_args = self.collect_type_const_args(&ast_generics.params);
+
+        let universal_args: Vec<_> = impl_trait_inputs.iter().map(|ty| match ty.kind {
+            TyKind::ImplTrait(node_id, _) => {
+                let def_id = self.local_def_id(node_id);
+                let span = self.lower_span(ty.span);
+                let ident = Ident::from_str_and_span(&pprust::ty_to_string(ty), span);
+                hir::GenericArg::Type(hir::Ty {
+                    hir_id: self.next_id(),
+                    span,
+                    kind: hir::TyKind::Path(hir::QPath::Resolved(None, self.arena.alloc(hir::Path {
+                        span,
+                        res: Res::Def(DefKind::TyParam, def_id.to_def_id()),
+                        segments: arena_vec![self; hir::PathSegment::from_ident(self.lower_ident(ident))],
+                    })))
+                })
+            }
+            _ => {
+                unreachable!(
+                    "impl_trait_inputs contains {:?} which is not TyKind::ImplTrait(..)",
+                    ty.kind
+                );
+            }
+        }).collect();
+
+        let generic_args = self.arena.alloc_from_iter(
+            lifetimes
+                .into_iter()
+                .chain(type_const_args.into_iter())
+                .chain(universal_args.into_iter()),
+        );
+
         // `impl Trait` now just becomes `Foo<'a, 'b, ..>`.
-        hir::TyKind::OpaqueDef(hir::ItemId { def_id: opaque_ty_def_id }, lifetimes)
+        let result = hir::TyKind::OpaqueDef(hir::ItemId { def_id: opaque_ty_def_id }, generic_args);
+        debug!("lower_opaque_impl_trait = {:?}", result);
+        result
```

Comment:
you can add `ret` to the `#[instrument]` on the function to generate this very code that you wrote by hand here